### ISSUE TITLE
Add in-game timer for speedruns

### ISSUE THREAD

**Issue by Gyoo** [renaudbedard/fez-1.12-issues#99]
Already talked about that several times on Discord, the speedrunning community would like to have an official in-game timer that counts the gameplay time without loading times, similar to what already exists in FEZmod

Basic info :
- Timer starts at the beginning of the animation when you start a new game, either from an empty or an existing file (so, either when you press "New game" or when you press OK on the confirmation to start a new game)
- Timer stops when you exit Gomez' room to either the 32 Cubes or the 64 Cubes ending village.

What happens in between is up to you. Either you just show the plain timer or you can add a few "splits" with milestones, for instance the number of cubes (every 8 cubes maybe ?)

To me, if a timer is implemented, there should be an option to enable/disable it, as well as the other speedrun-related features in case the speedrun "mode" is enabled.

**Comment by flibitijibibo** [comment]
We might be able to do something like Axiom Verge where we measure in frames, since the game always does logic updates at 60-ish Hz... 99% of the work is just getting it to time during level frames, which I wouldn't be much help with since I just do the FNA side of things.

Whether this goes into FEZ or FEZMod, some notes to consider from a design standpoint:
- HH:MM:SS:FF (F == Frames, always 17ms)
- Splits measured in cubes, since a speedrun's not going to bother with getting cubits that aren't going towards a cube
- Display the total time and the timestamp of the last cube earned during gameplay
- Display all cubes after hitting the ending village
- Previous timestamp data can be stored in a human-readable format, possibly just raw frames interpreted as 1-line-equals-1-cube, allows for record comparisons

Note that everything I've written is entirely speculative and doesn't factor in what it'd take to store this in the game, but someone who wants to write this might be able to make use of the UX prototype... if I missed any critical detail, just steal from Axiom Verge since Tom did that really well.

**Comment by Gyoo** [comment]
That's pretty much the idea we had in mind, and it was actually inspired by Axiom Verge.

I don't exactly know how it is managed in FEZmod (not sure it's done with frames), but the goal here is to have something directly implemented in the legacy game for more "officialness" and ease to use.

You have to know that loading times differ greatly from a computer to another, and even on my own and only computer, loading times are about 30/40 seconds longer (for a 28 minutes long run) when I stream my gameplay with OBS. Removing these loading times "in an official way" would be the best way to increase fairness between runners.

**Comment by flibitijibibo** [comment]
I've just implemented a prototype speedrun mode for the next update.

How it works is that you pass `--gotta-gomez-fast` as a launch parameter, and on startup there will be a "SPEEDRUN" option below the 3 saveslots. Select that and it will dump you right into a new game (which we store in saveslot "4" just in case). We add splits for every cube collected, and the time stops on the exact line of code where the ending cutscenes start.

The only thing that might be different from the normal mode is that I actually force singlethreaded mode when enabling speedrun mode. This can save a lot of time, even if the game stutters for slower machines. (Save the frames, kill the loading screens.)

Right now I'm just dumping text onto the screen, but I'm sure there's a way to pretty it up.

**Comment by flibitijibibo** [comment]
Final TODO before marking this done:
- ~~"Reset" option instead of "New Game"~~
- ~~Close run mode when new saveslot is used~~
- ~~Monospace font for printed times~~
- Pretty up text rendering (it's just one function, not as bad as you might think!)

The only real design change since the last post is that I now just give the user a single 'speedrun' option on startup instead of shoving it into the saveslot selection. I may backpedal on this depending on feedback, or if we sneak in a special speedrun image to render instead of a map preview in that space on the screen.

**Comment by flibitijibibo** [comment]
Just pushed a new build for Linux/Mac that should have this fully functioning now. Still need to make it pretty (and add locale strings), but it should work exactly as it will when this finally goes out.

**Comment by Gyoo** [comment]
I'll make sure to check it out when I have some time on my hands :)

**Comment by Gyoo** [comment]
Just tested real quickly to get a first look at your work.
It looks very nice so far (except for the [MISSING TEXT] in the menus :> )
You can see the timer well most of the time, except during the reboot cutscenes because of the white background. I only went to 4 cubes, but I assume all the cube splits are shown ? If so, it might get a bit cluttered near the end.

I'll do a full run soon to make sure everything is fine and all.

**Comment by flibitijibibo** [comment]
The next update will have a monospaced font that works reasonably well. But instead of a real font, it's just this image here:

![speedrun](https://cloud.githubusercontent.com/assets/1396814/15874535/89652aec-2cd1-11e6-90f3-b24034a0b83c.png)

For now you just throw this PNG into the game folder as "SpeedRun.png" and it'll load up (and will crash if you don't). We'll get the font into the pakfiles eventually.

Once we're all happy with how it looks we can mark this as done.

**Comment by flibitijibibo** [comment]
The next update should have the final menu/graphics for speedrun mode. Even when all 32 cubes are on screen it actually looks pretty good! Basically as long as it doesn't crash somehow this should be all set, but we can keep using this thread if anything particularly bad shows up.

**Comment by flibitijibibo** [comment]
The latest update adds dropshadows to the run text, which makes it appear in places like the fakereboot title screen. It even makes the cube characters look cool! I gotta remember this for future games.

**Comment by Gyoo** [comment]
Finally had time to make a full run and I've got some feedback for you ! Next to each point will be a timestamp, that refers to the following video : https://www.youtube.com/watch?v=P21YHJSTzE4
- On my screen resolution (Full screen 1920*1080), the splits don't show correctly. They're partially hidden at the right of the screen, and you can't see the times after the 30th cube because it goes below what you can see (29:19)
- In the infinite waterfall room (11:25) we make a cutscene skips that allows us to get 2 cubes with only one "Cube get" animation by spamming Jump fast enough. This leads to a cube not being count in the splits, so you they will show only 31 cubes instead of 32 : 
  [RUN_12062016_194729.txt](https://github.com/renaudbedard/fez-1.12-issues/files/310891/RUN_12062016_194729.txt)
- Please remove NG+ state of the Speedrun mode. I was really surprised to see that I had access to 3D view, and it actually prevented me from pulling off the hardest trick of the run because I triggered 3D view at the worst time :( (12:23)
- In the file that stores split times, can you add an \n after each split ? It's just much better looking :P

Other than that, I am really pleased with the way this is taking. Keep up the good work !

**Comment by flibitijibibo** [comment]
I'll have to do '\r'\n' for Windows, we have newlines but Notepad doesn't read them even in 2016...

I did a Ceiling at one point due to integer scaling of the text, I'll just replace that with a Floor.

I'll have to see if I can even track that dupe properly, I'm not sure how the addition happens exactly...

I think the only thing I _can't_ do is ignore NG+ states only because of how those modes are implemented... the speedrun mode is just a timer, it doesn't actually change gameplay. It might be skippable but it'll take more time than the others.

**Comment by Gyoo** [comment]
Alright. I understand if you can't remove NG+ from speedrun mode. We'll just have to be more careful :^)

**Comment by flibitijibibo** [comment]
Everything except the NG+ change will be in the next update. The downside to the cube cutscene skip fix is that the split is now explicitly when cubes are added to the count, so it will happen at the _end_ of the cutscene and not the beginning. It's more accurate but will probably feel weird at first.

**Comment by Gyoo** [comment]
I started a run fairly quickly to see how things go but I encountered something dramatically important that I should be ashamed not to have found beforehand so I didn't even continue the run to talk about it right now.

First, the good things :
- I like the smaller splits. It takes less space on the screen and all 32 cubes will fit easily.
- Having the split when the counter increases isn't disturbing imo. I know some people used to split there manually. Plus, if it fixes the previous issue, it's all good.

Now, THE thing that needs to be fixed.

Some loading times are not removed. I can see that sometimes the timer stops during small loading times, but for the bigger ones (especially the loading screens where you can see Dot in full screen), the timer keeps going when it shouldn't.

Basically, the 3 crucial loading zones are :
- Loading between rooms (with transition)
- Black screen loading (with the cube spinning in the bottom right)
- DOT loading screens (these are usually the longest ones, so from a computer to another, the difference can be big)

Oh, and one last cosmetic thing : Could you put the "SPEEDRUN" menu at the bottom of the save selection ? Makes more sense since it's more a bonus than anything else

**Comment by 0x0ade** [comment]
Not a dev, but my 2 cents:

# FEZMod did the loading check completely wrong.

_ahem_

The cases you listed additionally to the black screen are actually more transitions than loading screens (loading is hidden) - only the extra time spent loading should be ignored. It's just that the transition itself already triggers the loading flag.

In my tests, the faraway transition cut to the black loading screen when loading took longer than the transition. The Dot transition should just take longer, but it functions as palate cleanser anyway.

**Comment by Gyoo** [comment]
Well, from what you're telling here, maybe I need to make sure what a loading really is.

I know for fact that, on the 3 elements I've listed, the time can differ from a computer to another. Now, does the whole animations count as loading or just part of them (which would mean, there is a minimal "loading" time that is the same for everyone), this is where I lack data.

**Comment by flibitijibibo** [comment]
The speedrun mode always runs in singlethreaded mode, so times should now be identical across machines above a certain spec. The boxes that hitch at loading points may be a few ms slower, but I've not found a setup that does this yet.

The only screen I'm not 100% sure on is the dot screen. The async loading screens, like the bottom right cube screen, won't happen because the game loads too fast in singlethreaded mode.

**Comment by Gyoo** [comment]
On my last run, I notcied that there was no difference between my Real Time timer (using Livesplit) and the In Game Timer.
What we noticed with @renaudbedard is that, most likely, the timer freezes during loads, but actually doesn't cut off loading times. Maybe he'll be able to explain it better than me as we witnessed this together on my stream.

**Comment by flibitijibibo** [comment]
I'm surprised there are any loading times... I can see if there's a way to pause the timer when the engine claims to be loading, but I can't guarantee that the spot that hangs will be exactly when the timer stops; could be a couple microseconds off.

**Comment by Gyoo** [comment]
If it's only microseconds, we can bear with it :)

**Comment by flibitijibibo** [comment]
I've pushed a change that adds pauses/resumes for level changes specifically at the points where level loads start and when loading is officially declared finished in the engine. This does indeed shave off a fraction of a second per room, and for the Dot loading screen, you get the added bonus of having the sub-millisecond time to look at while you wait for it to load :P

I think what I wrote covers all scenarios, but Renaud may have better places to put my new timer calls. We'll see in the next build.

**Comment by Gyoo** [comment]
I'll make sure to test it asap :)

**Comment by Gyoo** [comment]
Alright, just did another run to test the changes you did to the in-game timer.

Now, what the game considers a loading is being cut off correctly from the time, but I've noticed 2 things at the end of the run, one is less important than the other :
- The cutscene at the top of gravity room (with the hexahedron being filled with the cubes you gathered) is not counted in the time. This could actually make sense since this is not a playable moment and there's nothing you can do there apart from watching. However, if possible, the timer should restart when you're able to press A after the fake reset, because that's an input you can lose time on if you don't pay attention :)
- After you cross Gomez' room's door at the end of the run, the timer doesn't stop as it used to do.

Please don't kill me :(

**Comment by flibitijibibo** [comment]
Both are a consequence of the loading time fixes - the game starts the cutscene while it's in the load process, so at the end of load it resumes the time even after we're done!

The gravity room cutscene, I've no idea - this may be where I have to defer to Renaud, as I'm not entirely sure what the rules are for when the game is actually loading and when the game is just changing levels. We have an IsLoading check internally, but it's not what you might expect it to mean.

For now I'm just going to revert my latest speedrun change since I'd rather incorporate load times and add a few ms to times rather than be whole minutes off and not have an ending >_>

**Comment by renaudbedard** [comment]
Reopening so I can take a look later.

**Comment by renaudbedard** [comment]
Fixed! I did a spot check for "dot loadings", normal level transitions, the beginning cutscenes and the 32-bit end cutscene, everything seemed good. Load times no longer are included in the time counter and it seems to behave like expected everywhere.

**Comment by Gyoo** [comment]
One last check : what about the 64 cubes ending ? :)

**Comment by flibitijibibo** [comment]
Looks like the 64-cube ending should work as well, as the new timer pauses are based only on load times (and this time it actually checks to see if time was already called before resuming!).
